### PR TITLE
New feature: set a note as favorite (star/unstar)

### LINKED ISSUE #202: Star/Unstar
Starred Notes should be displayed like in a folder  at the first position.

Refers to #2

### PR DESCRIPTION
**Pull request by korelstar**
This pull request brings a star/unstar-feature for notes as known already from other owncloud apps: you can click on a gray star to mark a note as your favorite, then the note is shown on the top of the list and a yellow star appears, so you can easily find your most important notes.

In the technical backend, it uses the existing system to star files, i.e. a special tag for the file. The frontend provides respective new actions.

closes #202 

### (Open) tasks
- [x] provide setting/unsetting notes as favorite (star/unstar)
- [x] extend API for 3rd-party apps

### CODE REVIEW

**Comment by korelstar** [comment]
For the API, there are multiple approaches for providing this feature to 3rd-party apps, e.g.:

1. Introduce a new optional field `favorite` for creating/updating nodes, beside the existing `content`. Since `modified` should not be changed by starring a note, `content` have to be made optional. Then, `modified` could be updated if and only if a `content` is given, otherwise only `favorite` will be updated.
2. Introduce a new route for meta-data of a note. In the future, this could also include categories or tags. Here, it is clear, that `modified` will not be updated.

What do you think?

File: css/notes.css [hunk 1]
```diff
@@ -43,9 +43,17 @@
...
 
 #app-navigation .active > .utils .icon-delete {
-    display: block;
+    display: inline-block;
     opacity: .3;
 }
+#app-navigation .active > .utils .icon-star {
+    display: inline-block;
+    opacity: .3;
+}
```
Comment (Henni) [inline css/notes.css]: please merge this rule with the previous rule:
```css
#app-navigation .active > .utils .icon-delete,
#app-navigation .active > .utils .icon-star {
...
}
```

File: css/notes.css [hunk 2]
```diff
@@ -43,9 +43,17 @@
...
+#app-navigation .utils .icon-starred {
+    display: inline-block;
+    opacity: 1 !important;
+}
```
Comment (Henni) [inline css/notes.css]: It should be possible to remove the `!important` if the rule is set to:
```css
#app-navigation .active > .utils .icon-starred {
```

File: js/app/services/notesmodel.js [hunk 3]
```diff
@@ -49,8 +50,8 @@ app.factory('NotesModel', function () {
                     break;
                 }
             }
-        }
+        },
```
Comment (Henni) [inline js/app/services/notesmodel.js]: Where does this `,` come from?

File: css/notes.css (hunk 1 above)
Comment (korelstar) [inline css/notes.css]: done

File: js/app/services/notesmodel.js (hunk 3 above)
Comment (korelstar) [inline js/app/services/notesmodel.js]: copy&paste, sorry. It's removed now.

File: db/note.php [hunk 4]
```diff
@@ -24,31 +24,38 @@
...
         $this->addType('modified', 'integer');
+        $this->addType('favorite', 'boolean');
     }
 
     /**
      * @param File $file
      * @return static
      */
-    public static function fromFile(File $file){
+    public static function fromFile(File $file, array $tags=[]){
```
Comment (Henni) [inline db/note.php]: If there are no favorites, the `array` annotation results in an Internal Server Error.

image

Comment (Henni) [inline db/note.php]: BTW: This is also why travis fails.

**Comment by korelstar** [comment]
Another thing: The automated tests failed for my pull request, because `\OC::$server->getTagManager()->load('files')` returns `null`. However, as far as I know, this is the correct official approach to get an instance of `OCP\ITags` (which is also used in owncloud/core: e.g., [Application](https://github.com/owncloud/core/blob/37177dd54d67443cacc0061c17f95070442b93cf/apps/files/lib/AppInfo/Application.php#L79), [Helper](https://github.com/owncloud/core/blob/37177dd54d67443cacc0061c17f95070442b93cf/apps/files/lib/Helper.php#L217), [TagServiceTest](https://github.com/owncloud/core/blob/37177dd54d67443cacc0061c17f95070442b93cf/apps/files/tests/Service/TagServiceTest.php#L76)) and it works on my installation.

Therefore, I believe that the test environment is not sufficient. However, I have no idea how to change it, so that the tests will be successful. Can anyone help please?

**Comment by jancborchardt** [comment]
Really awesome work @korelstar! :) Great that it directly hooks in with Files, it’s just perfect.

File: db/note.php [hunk 5]
```diff
@@ -44,7 +44,7 @@ public function __construct() {
      * @param File $file
      * @return static
      */
-    public static function fromFile(File $file, array $tags=[]){
+    public static function fromFile(File $file, $tags=[]){
```
Comment (Henni) [inline db/note.php]: as soon as travis is happy with this change, I'll merge it.

**Comment by Henni** [comment]
@korelstar I haven't figured out why travis fails. Maybe you can figure it out:
```
Call to a member function getTagsForObjects() on a non-object in [...]/apps/notes/service/notesservice.php on line 54
```
https://travis-ci.org/owncloud/notes/jobs/169027138

**Comment by Henni** [comment]
`$tagger` might be `null` if no user is logged in.

**Comment by korelstar** [comment]
Which wording should be used? I've chosen `favorite`, because this term is used by the existing tagging system. However, `star/unstar` is more neutral and leaves space for other interpretations (e.g, *important*, *to be done*, *interesting*), so I like that wording, too. I think, I prefer this one.

What do you think? I will adjust my changes accordingly.

@jancborchardt @Henni

**Comment by Henni** [comment]
@korelstar I'm happy with both. @jancborchardt what do you prefer?

**Comment by Henni** [comment]
Two things that should be done:
- [ ] Inject TagManager instead of using \OC::$server->getTagManager() (notes uses [automatic dependency assembly](https://doc.owncloud.org/server/9.1/developer_manual/app/container.html?highlight=injection#use-automatic-dependency-assembly-recommended))
- [ ] Extend tests (requires dependency injection)

I'm also willing to do this. But then I would merge this PR first and fix these two points in a separate PR. (Working on two different repos is painful)

**Comment by jancborchardt** [comment]
Let's go with »Favorite« for now since it's symmetric to how we use it elsewhere. :)

**Comment by korelstar** [comment]
@Henni 
That would be nice, if you could do the two tasks.
A few minutes ago, I've added the possibility to set favorites through the API (I used variant 1 of my post above: https://github.com/owncloud/notes/pull/248#issuecomment-254061091). Hence, I'm ready for merge, now! (however, my commit did not trigger Travis)

**Comment by Henni** [comment]
Travis finally has run through.
So 👍
